### PR TITLE
pkg/uzlib: add support for zlib compression/decompression

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR adds support for zlib compression/decompression using a package from https://github.com/pfalcon/uzlib.

### Testing procedure

- Green CI
- Test application works:

<details><summary>native</summary>

```
$ BUILD_IN_DOCKER=1 make -C tests/pkg_uzlib all test --no-print-directory 
/work/riot/RIOT/makefiles/kconfig.mk:105: Warning! SHOULD_RUN_KCONFIG is not set but a previous configuration file was detected (did you run `make menuconfig`?). Kconfig will run regardless.
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'        -w '/data/riotbuild/riotbase/tests/pkg_uzlib/' 'riot/riotbuild:latest' make     all
/data/riotbuild/riotbase/makefiles/kconfig.mk:105: Warning! SHOULD_RUN_KCONFIG is not set but a previous configuration file was detected (did you run `make menuconfig`?). Kconfig will run regardless.
Building application "tests_pkg_uzlib" for "native" with MCU "native".

"make" -C /data/riotbuild/riotbase/pkg/uzlib
"make" -C /data/riotbuild/riotbase/build/pkg/uzlib/src -f /data/riotbuild/riotbase/pkg/uzlib/uzlib.mk
"make" -C /data/riotbuild/riotbase/boards/native
"make" -C /data/riotbuild/riotbase/boards/native/drivers
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/native
"make" -C /data/riotbuild/riotbase/cpu/native/periph
"make" -C /data/riotbuild/riotbase/cpu/native/stdio_native
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/test_utils/interactive_sync
   text	   data	    bss	    dec	    hex	filename
  36334	    632	  40580	  77546	  12eea	/data/riotbuild/riotbase/tests/pkg_uzlib/bin/native/tests_pkg_uzlib.elf
r
/work/riot/RIOT/tests/pkg_uzlib/bin/native/tests_pkg_uzlib.elf /dev/ttyACM0 
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2022.01-devel-456-g33d20-pr/pkg/uzlib)
Compressing message (len: 883): "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nisl vel pretium lectus quam id leo. Volutpat sed cras ornare arcu dui vivamus arcu felis bibendum. Velit scelerisque in dictum non consectetur a erat nam. Pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus. Tristique et egestas quis ipsum suspendisse. At imperdiet dui accumsan sit amet nulla facilisi. Pulvinar neque laoreet suspendisse interdum consectetur libero. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae. Sed adipiscing diam donec adipiscing tristique risus nec. Venenatis tellus in metus vulputate eu scelerisque. Id faucibus nisl tincidunt eget nullam non nisi est. Integer feugiat scelerisque varius morbi enim. Est sit amet facilisis magna etiam. Venenatis cras sed felis eget velit aliquet sagittis."
Message compressed to 813 raw bytes (ratio: 0.92)
Decompressed message (len: 883): "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nisl vel pretium lectus quam id leo. Volutpat sed cras ornare arcu dui vivamus arcu felis bibendum. Velit scelerisque in dictum non consectetur a erat nam. Pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus. Tristique et egestas quis ipsum suspendisse. At imperdiet dui accumsan sit amet nulla facilisi. Pulvinar neque laoreet suspendisse interdum consectetur libero. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae. Sed adipiscing diam donec adipiscing tristique risus nec. Venenatis tellus in metus vulputate eu scelerisque. Id faucibus nisl tincidunt eget nullam non nisi est. Integer feugiat scelerisque varius morbi enim. Est sit amet facilisis magna etiam. Venenatis cras sed felis eget velit aliquet sagittis."

```

</details>

<details><summary>nrf52832-mdk</summary>

```
$ BUILD_IN_DOCKER=1 BOARD=nrf52832-mdk make -C tests/pkg_uzlib flash test --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=nrf52832-mdk'  -w '/data/riotbuild/riotbase/tests/pkg_uzlib/' 'riot/riotbuild:latest' make     
Building application "tests_pkg_uzlib" for "nrf52832-mdk" with MCU "nrf52".

"make" -C /data/riotbuild/riotbase/pkg/uzlib
"make" -C /data/riotbuild/riotbase/build/pkg/uzlib/src -f /data/riotbuild/riotbase/pkg/uzlib/uzlib.mk
"make" -C /data/riotbuild/riotbase/boards/nrf52832-mdk
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/nrf52
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/vectors
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common/periph
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/sys/malloc_thread_safe
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
"make" -C /data/riotbuild/riotbase/sys/test_utils/interactive_sync
"make" -C /data/riotbuild/riotbase/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  16348	    128	   5972	  22448	   57b0	/data/riotbuild/riotbase/tests/pkg_uzlib/bin/nrf52832-mdk/tests_pkg_uzlib.elf
/work/riot/RIOT/dist/tools/pyocd/pyocd.sh flash /work/riot/RIOT/tests/pkg_uzlib/bin/nrf52832-mdk/tests_pkg_uzlib.hex
### Flashing Target ###
0004678:INFO:load_cmd:Loading /work/riot/RIOT/tests/pkg_uzlib/bin/nrf52832-mdk/tests_pkg_uzlib.hex at 0x00000000
[==================================================] 100%
0006807:INFO:loader:Erased 20480 bytes (5 sectors), programmed 20480 bytes (5 pages), skipped 0 bytes (0 pages) at 9.45 kB/s
Done flashing
r
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2022.01-devel-456-g33d20-pr/pkg/uzlib)
Compressing message (len: 883): "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nisl vel pretium lectus quam id leo. Volutpat sed cras ornare arcu dui vivamus arcu felis bibendum. Velit scelerisque in dictum non consectetur a erat nam. Pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus. Tristique et egestas quis ipsum suspendisse. At imperdiet dui accumsan sit amet nulla facilisi. Pulvinar neque laoreet suspendisse interdum consectetur libero. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae. Sed adipiscing diam donec adipiscing tristique risus nec. Venenatis tellus in metus vulputate eu scelerisque. Id faucibus nisl tincidunt eget nullam non nisi est. Integer feugiat scelerisque varius morbi enim. Est sit amet facilisis magna etiam. Venenatis cras sed felis eget velit aliquet sagittis."
Message compressed to 813 raw bytes (ratio: 0.92)
Decompressed message (len: 883): "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Nisl vel pretium lectus quam id leo. Volutpat sed cras ornare arcu dui vivamus arcu felis bibendum. Velit scelerisque in dictum non consectetur a erat nam. Pretium viverra suspendisse potenti nullam ac tortor vitae purus faucibus. Tristique et egestas quis ipsum suspendisse. At imperdiet dui accumsan sit amet nulla facilisi. Pulvinar neque laoreet suspendisse interdum consectetur libero. Vulputate sapien nec sagittis aliquam malesuada bibendum arcu vitae. Sed adipiscing diam donec adipiscing tristique risus nec. Venenatis tellus in metus vulputate eu scelerisque. Id faucibus nisl tincidunt eget nullam non nisi est. Integer feugiat scelerisque varius morbi enim. Est sit amet facilisis magna etiam. Venenatis cras sed felis eget velit aliquet sagittis."

```

</details>

### Issues/PRs references

None

### CODE REVIEW

File: tests/pkg_uzlib/main.c [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    struct uzlib_comp comp = {0};
+    comp.dict_size = 32768;
+    comp.hash_bits = 6;
+    comp.hash_table = hash_table;
+
+    zlib_start_block(&comp.out);
+    uzlib_compress(&comp, (uint8_t *)message, strlen(message));
+    zlib_finish_block(&comp.out);
+
+    float compression_ratio = (comp.out.outlen * 100.0) / strlen(message);
```
Comment (benpicco) [inline tests/pkg_uzlib/main.c:73]: Why float?

```suggestion
    unsigned compression_ratio = (comp.out.outlen * 10000) / strlen(message);
```

File: tests/pkg_uzlib/main.c [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    printf("Compressing message (len: %u): \"%s\"\n",
+           strlen(message), message);
+    printf("Message compressed to %u raw bytes (ratio: %i.%i)\n",
+           comp.out.outlen,
+           (int)compression_ratio / 100, (int)compression_ratio % 100) ;
```
Comment (benpicco) [inline tests/pkg_uzlib/main.c:79]:
```suggestion
           compression_ratio / 10000, (compression_ratio / 100) % 100);
```

**Review by benpicco — COMMENTED** [review]
There are a few more static checks

File: tests/pkg_uzlib/main.c [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    /* Decompress */
+    uzlib_init();
+    unsigned int dlen;
+
+    /* Get decompressed length */
+    dlen =            comp.out.outbuf[comp.out.outlen - 1];
+    dlen = 256 * dlen + comp.out.outbuf[comp.out.outlen - 2];
+    dlen = 256 * dlen + comp.out.outbuf[comp.out.outlen - 3];
+    dlen = 256 * dlen + comp.out.outbuf[comp.out.outlen - 4];
```
Comment (benpicco) [inline tests/pkg_uzlib/main.c:89]:
```suggestion
    dlen = unaligned_get_u32(&comp.out.outbuf[comp.out.outlen - 4]);
```

If that's not in host byte order, there is `byteorder_bebuftohl()`

File: tests/pkg_uzlib/main.c [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    struct uzlib_uncomp d;
```
Comment (benpicco) [inline tests/pkg_uzlib/main.c:91]:
```suggestion
    struct uzlib_uncomp d = {
        .source = comp.out.outbuf,
        .source_limit = comp.out.outbuf + comp.out.outlen - 1,
        .dest_start = d.dest = decompressed_output,
    }
```

does this work?

File: tests/pkg_uzlib/main.c (hunk 3 above)
Comment (aabadie) [reply tests/pkg_uzlib/main.c]: Thanks, that's more concise.

File: tests/pkg_uzlib/main.c (hunk 4 above)
Comment (aabadie) [reply tests/pkg_uzlib/main.c]: It works, so I modified it this way

**Comment by aabadie** [comment]
@benpicco this is all green, care to ACK ?

**Review by benpicco — APPROVED** [review]

**Pull request merged**